Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTP1StateMachine.swift

```diff
@@ -0,0 +1,327 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+extension HTTPConnectionPool {
+    struct HTTP1StateMachine {
+        enum State: Equatable {
+            case running
+            case shuttingDown(unclean: Bool)
+            case shutDown
+        }
+
+        typealias Action = HTTPConnectionPool.StateMachine.Action
+
+        private var connections: HTTP1Connections
+        private var failedConsecutiveConnectionAttempts: Int = 0
+
+        private var queue: RequestQueue
+        private var state: State = .running
+
+        init(idGenerator: Connection.ID.Generator, maximumConcurrentConnections: Int) {
+            self.connections = HTTP1Connections(
+                maximumConcurrentConnections: maximumConcurrentConnections,
+                generator: idGenerator
+            )
+
+            self.queue = RequestQueue()
+        }
+
+        // MARK: - Events -
+
+        mutating func executeRequest(
+            _ request: HTTPSchedulableRequest,
+            onPreferred preferredEL: EventLoop,
+            required: Bool
+        ) -> Action {
+            switch self.state {
+            case .running:
+                break
+            case .shuttingDown, .shutDown:
+                // it is fairly unlikely that this condition is met, since the ConnectionPoolManager
+                // also fails new requests immediately, if it is shutting down. However there might
+                // be race conditions in which a request passes through a running connection pool
+                // manager, but hits a connection pool that is already shutting down.
+                //
+                // (Order in one lock does not guarantee order in the next lock!)
+                return .init(.failRequest(request, HTTPClientError.alreadyShutdown, cancelWaiter: nil), .none)
+            }
+
+            if required {
+                preconditionFailure("EventLoop requirements not supported as of today.")
```

Comment:
I feel like it's a little pointless to have `required` as a parameter if we can't do anything with it